### PR TITLE
Migrate to django-environ

### PR DESCRIPTION
**Pull request by tkdchen**
This patch uses django-environ and drop deprecated and old packages
django-cache-url, dj-email-url, dj-database-url and dj-cache-url.

Tests are updated according to the behavior of django-environ.

Signed-off-by: Chenxiong Qi <qcxhome@gmail.com>

### CODE REVIEW

File: tests/test_values.py [hunk 1]
```diff
@@ -443,13 +439,7 @@ def test_cache_url_value(self):
...
             with self.assertRaises(Exception) as cm:
                 value.setup('TEST')
-            self.assertEqual(cm.exception.args[0], 'Unknown backend: "wrong"')
-        with env(CACHE_URL='redis://user@host:port/1'):
-            with self.assertRaises(ValueError) as cm:
-                value.setup('TEST')
-            self.assertEqual(
-                cm.exception.args[0],
-                "Cannot interpret cache URL value 'redis://user@host:port/1'")
+            self.assertEqual(cm.exception.args[0], 'wrong')
```
Comment (blueyed) [inline tests/test_values.py]: That looks "wrong"?!
Shouldn't it have more info / a full msg?

File: tests/test_values.py [hunk 2]
```diff
@@ -411,26 +410,23 @@ def test_email_url_value(self):
...
+                'EMAIL_PORT': None})
         with env(EMAIL_URL='smtps://user@domain.com:password@smtp.example.com:wrong'):  # noqa: E501
             self.assertRaises(ValueError, value.setup, 'TEST')
 
     def test_cache_url_value(self):
         cache_setting = {
             'default': {
                 'BACKEND': 'django_redis.cache.RedisCache',
-                'LOCATION': 'redis://host:6379/1',
+                'LOCATION': 'redis://user@host:6379/1',
```
Comment (blueyed) [inline tests/test_values.py]: Apparently it changes some defaults (more secure by default).

File: tests/test_values.py [hunk 3]
```diff
@@ -411,26 +410,23 @@ def test_email_url_value(self):
                 'EMAIL_HOST_PASSWORD': 'password',
                 'EMAIL_HOST_USER': 'user@domain.com',
                 'EMAIL_PORT': 587,
-                'EMAIL_USE_SSL': False,
                 'EMAIL_USE_TLS': True})
-        with env(EMAIL_URL='console://'):
+        with env(EMAIL_URL='consolemail://'):
```
Comment (blueyed) [inline tests/test_values.py]: Is that change required?
Is it related to https://github.com/psgs/ConsoleMail?

File: setup.py [hunk 4]
```diff
@@ -41,18 +41,11 @@ def find_version(*parts):
...
         'testing': [
             'django-discover-runner',
             'mock',
-            'django-cache-url>=1.0.0',
-            'dj-database-url',
-            'dj-email-url',
-            'dj-search-url',
             'six',
             'Sphinx>=1.4',
         ],
```
Comment (blueyed) [inline setup.py]: Could maybe still be in `extras_require`?

File: tests/test_values.py (hunk 3 above)
Comment (tkdchen) [reply tests/test_values.py]: > Is it related to https://github.com/psgs/ConsoleMail?

No. I'm not aware of ConsoleMail. What I learned from django-environ is only `consolemail` is defined in email schemes, but no `console`. The `consolemail` maps to same backend `django.core.mail.backends.console.EmailBackend`.

File: tests/test_values.py (hunk 1 above)
Comment (tkdchen) [reply tests/test_values.py]: django-environ simply raises `KeyError` when looking up a cache scheme by `wrong`, and here `assertRaises` catches the `KeyError`.

It might be good for django-envion to catch the KeyError and reraise a specific error with more descriptive message.

What do you think?

File: tests/test_values.py (hunk 2 above)
Comment (tkdchen) [reply tests/test_values.py]: I'm not sure if I get your point. Is this change acceptable?

Comment (blueyed) [reply tests/test_values.py]: It changes defaults apparently, so we should bump the major or at least minor version for it.

File: tests/test_values.py (hunk 3 above)
Comment (blueyed) [reply tests/test_values.py]: That's another behavior change then apparently AFAIK.

All those might be fine, and it's good to modernize, but might break setups/configs.

File: tests/test_values.py (hunk 1 above)
Comment (tkdchen) [reply tests/test_values.py]: FYI, I made a patch to django-environ, https://github.com/joke2k/django-environ/pull/235

File: setup.py (hunk 4 above)
Comment (tkdchen) [reply setup.py]: Looks no need here. How useful to keep those in extra_require?

Comment (blueyed) [reply setup.py]: It avoids having to install / installing it by default when not needed.

I guess the typical usecase is using URLs for configs then, so that seems to be OK.

File: tests/test_values.py (hunk 3 above)
Comment (tkdchen) [reply tests/test_values.py]: Hi @blueyed, I added another commit to make `console://` still work in apps using it, and deprecate `console://`.

**Comment by pauloxnet** [comment]
I think `django-environ` is a strong addiction. I would prefer to leave `django-configuration` as light as possible.

**Comment by carltongibson** [comment]
Hi folks. 

Can you see also https://github.com/django/django/pull/13029 and the [related mailing list discussion](https://groups.google.com/d/topic/django-developers/CIPgeTetYpk/discussion)? 

**tl;dr**: Maybe we pull something into Django here. Perhaps it makes sense to have env vars read by the default project templates, but we should get right exactly what that should look like, and I don't want us to reinvent something that's already _solved™_ by the community. 

It would be good to get your thoughts as contributors here. 

Thanks 😉
C.

File: tests/test_values.py (hunk 1 above)
Comment (jayvdb) [reply tests/test_values.py:446]: @tkdchen , that patch was merged; I guess this can be updated now ?

**Review by auvipy — CHANGES_REQUESTED** [review]
I am very positive about this change moves o django-environ. can we just rebase and get this merged

**Review by brianhelba — CHANGES_REQUESTED** [review]
@auvipy There are a some issues with your recent merge. See inline comments.

**Review by brianhelba — CHANGES_REQUESTED** [review]
The docs for these Value types need to be updated too, as they contain references to the old upstream packages: https://github.com/jazzband/django-configurations/blob/master/docs/values.rst#url-based-values

Users looking for precise syntax info should be pointed to `django-environ` docs instead.

**Comment by uhurusurfa** [comment]
I would agree with @pauloxnet on this one and would go a step further by creating an internal URL checker and converter to remove the dependencies on the 3 external packages. I had to fix a unit test that started failing due to a change in dj_database_url
I would propose closing this PR unless there remains strong opinions to move forward with the django-environ use.
